File: js/39.combination-sum.js
/*
 * @lc app=leetcode id=39 lang=javascript
 *
 * [39] Combination Sum
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
	let res = [];
	let combination = [];

	dfs(candidates, target, 0, combination, res);
	console.log(res);
	return res;

	function dfs(candidates, target, start, combination, res) {
		// console.log('candidates, target, start, combination, res');
        // console.log(candidates, target, start, combination, res);
        
		if (target === 0) {
			// deep copy
			res.push(Array.from(combination));
			return;
		}
		if (target < 0) {
			return;
		}

		for (let i = 0; i < candidates.length; i++) {
			combination.push(candidates[i]);
			// instead of writing this

			// target -= candidates[i];
			// dfs(candidates, target, i, combination, res);
			// target += candidates[i];

			//you could write this
			dfs(candidates.slice(i), target - candidates[i], i, combination, res);
            combination.pop();
            
            // // it is not good, as the destructure combination use additional O(n)
            // dfs(candidates.slice(i), target - candidates[i], i, [...combination, candidates[i]], res);


		}
	}
};
// @lc code=end

File: 39.combination-sum.java
import java.util.*;

/*
 * @lc app=leetcode id=39 lang=java
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(candidates, 0, new ArrayList<>(), target, res);
        return res;
    }

    private void dfs(int[] candidates, int p, List<Integer> path, int rest, List<List<Integer>> res) {
        if(rest == 0){
            res.add(new ArrayList<Integer>(path));
        }
        for(int i=p; i<candidates.length; i++){
            if(rest - candidates[i] >= 0 ){
                path.add(candidates[i]);
                dfs(candidates, i, path, rest-candidates[i], res);
                path.remove(path.size()-1);
            }
        }
    }

}
// @lc code=end

